consensus/replay.go

```diff
@@ -278,6 +287,53 @@ func (h *Handshaker) Handshake(proxyApp proxy.AppConns) error {
 	return nil
 }
 
+// LocalSync restores stateStore and blockStore state which is not included in the snapshot
+// This function assumes that the implicit trusted local snapshot chunks have already been applied
+// State restoration depends on heights after the restored snapshot so the RPC dependency remains
+func (h *Handshaker) localSync(appBlockHeight uint64) (sm.State, *types.Commit, error) {
+	var state sm.State
+	var commit *types.Commit
+	var err error
+
+	if h.stateProvider == nil {
+		return sm.State{}, nil, errors.New("Missing State Provider")
+	}
+
+	pctx, pcancel := context.WithTimeout(context.TODO(), 30*time.Second)
+	defer pcancel()
+
+	// Optimistically build new state, so we don't discover any light client failures at the end.
+	state, err = h.stateProvider.State(pctx, appBlockHeight)
+	if err != nil {
+		h.logger.Error("failed to fetch and verify tendermint state", "err", err)
+		if err == light.ErrNoWitnesses {
+			return sm.State{}, nil, err
+		}
+		return sm.State{}, nil, errors.New("snapshot was rejected")
+	}
+
+	commit, err = h.stateProvider.Commit(pctx, appBlockHeight)
+	if err != nil {
+		h.logger.Error("failed to fetch and verify commit", "err", err)
+		if err == light.ErrNoWitnesses {
+			return sm.State{}, nil, err
+		}
+		return sm.State{}, nil, errors.New("snapshot was rejected")
+	}
+
+	if err := h.stateStore.Save(state); err != nil {
+		return sm.State{}, nil, err
+	}
+	if err := h.store.(*store.BlockStore).SaveSeenCommit(int64(appBlockHeight), commit); err != nil {
```

Comment:
Following the original implementation, which is `Node.startStateSync` (I don't know the branch to refer, so no link, sorry), you should use `state.LastBlockHeight` here for compatibility. It should be equal to `appBlockHeight`, except for uint/int stuff, but for coherence it is better to follow the same approach.